File: packages/browser/src/blob-asset.ts
import { decode } from "@msgpack/msgpack";
import { portableJsonObject } from "@marimo-team/portable-json";

import { parseMediaType } from "./media-type.js";
import { canonicalJson } from "./schema.js";
import { validateCanonicalMessagePack } from "./strict-msgpack.js";
import type { BlobAsset, BlobAssetDescriptor } from "./types.js";
import { isNotebookExportError, NotebookExportError } from "./types.js";
import { isRecordValue, isStringValue } from "./value-types.js";

const EXPECTED_FIELDS = ["data", "media_type", "filename", "metadata"];

export function decodeBlobAsset(bytes: Uint8Array, descriptor: BlobAssetDescriptor): BlobAsset {
  try {
    validateCanonicalMessagePack(bytes);
    const decoded = decode(bytes);
    if (!isRecordValue(decoded)) {
      throw new TypeError("BlobAsset must be a map.");
    }
    const fields = Object.keys(decoded);
    if (
      fields.length !== EXPECTED_FIELDS.length ||
      fields.some((field, index) => field !== EXPECTED_FIELDS[index])
    ) {
      throw new TypeError("BlobAsset must use the native four-field envelope.");
    }
    if (!("data" in decoded) || !(decoded.data instanceof Uint8Array)) {
      throw new TypeError("BlobAsset.data must be binary.");
    }
    if (!("media_type" in decoded) || !isStringValue(decoded.media_type)) {
      throw new TypeError("BlobAsset.media_type must be a string.");
    }
    const mediaType = parseMediaType(decoded.media_type);
    if (!("filename" in decoded)) throw new TypeError("BlobAsset.filename must be present.");
    const filename = decoded.filename;
    if (filename !== null && !isStringValue(filename)) {
      throw new TypeError("BlobAsset.filename must be a string or null.");
    }
    if (!("metadata" in decoded)) throw new TypeError("BlobAsset.metadata must be present.");
    const metadata = portableJsonObject(decoded.metadata, "BlobAsset.metadata");
    if (decoded.media_type !== descriptor.mediaType) {
      throw new TypeError("BlobAsset.media_type does not match its descriptor.");
    }
    if (filename !== descriptor.filename) {
      throw new TypeError("BlobAsset.filename does not match its descriptor.");
    }
    if (canonicalJson(metadata) !== canonicalJson(descriptor.metadata)) {
      throw new TypeError("BlobAsset.metadata does not match its descriptor.");
    }
    return Object.freeze({
      data: decoded.data,
      mediaType,
      filename,
      metadata,
    });
  } catch (error) {
    if (isNotebookExportError(error) && error.code === "asset_invalid") throw error;
    throw new NotebookExportError("asset_invalid", "BlobAsset envelope validation failed.", {
      cause: error,
      details: { outputCodec: descriptor.codec, mediaType: descriptor.mediaType },
    });
  }
}

File: packages/python/src/marimo_export/_blob_asset.py
from __future__ import annotations

from dataclasses import dataclass
from typing import cast

import msgspec

from marimo_export._format import MAX_FORMAT_METADATA_JSON_BYTES, validate_media_type
from marimo_export._json import JsonObject, canonical_bytes, portable_json_object
from marimo_export._portable import validate_portable_basename

_FIELDS = ("data", "media_type", "filename", "metadata")


@dataclass(frozen=True, slots=True)
class BlobAssetEnvelope:
    """The verified native marimo BlobAsset MessagePack envelope."""

    data: bytes
    media_type: str | None
    filename: str | None
    metadata: JsonObject


def decode_blob_asset(
    data: bytes,
    *,
    maximum_bytes: int | None = None,
) -> BlobAssetEnvelope:
    """Strictly decode the exact MessagePack shape emitted by marimo."""

    if not isinstance(data, bytes):
        raise TypeError("BlobAsset MessagePack must be bytes")
    if maximum_bytes is not None and (
        not isinstance(maximum_bytes, int) or isinstance(maximum_bytes, bool) or maximum_bytes <= 0
    ):
        raise TypeError("maximum_bytes must be a positive integer")
    if maximum_bytes is not None and len(data) > maximum_bytes:
        raise ValueError(f"BlobAsset MessagePack exceeds the {maximum_bytes} byte limit")

    try:
        decoded = msgspec.msgpack.decode(data)
    except msgspec.DecodeError as error:
        raise ValueError("BlobAsset is invalid MessagePack") from error
    if not isinstance(decoded, dict) or tuple(decoded) != _FIELDS:
        raise ValueError("BlobAsset does not use the native four-field envelope")
    if msgspec.msgpack.encode(decoded) != data:
        raise ValueError("BlobAsset MessagePack is not canonical")

    payload = decoded["data"]
    media_type = decoded["media_type"]
    filename = decoded["filename"]
    metadata = decoded["metadata"]
    if not isinstance(payload, bytes):
        raise ValueError("BlobAsset data must be bytes")
    if media_type is not None:
        try:
            media_type = validate_media_type(media_type, "BlobAsset media_type")
        except (TypeError, ValueError) as error:
            raise ValueError("BlobAsset media_type is invalid") from error
    if filename is not None:
        try:
            filename = validate_portable_basename(filename, "BlobAsset filename")
        except (TypeError, ValueError) as error:
            raise ValueError("BlobAsset filename is invalid") from error
    try:
        metadata_object = portable_json_object(metadata, "BlobAsset metadata")
    except (TypeError, ValueError) as error:
        raise ValueError("BlobAsset metadata must be portable JSON") from error
    if len(canonical_bytes(metadata_object)) > MAX_FORMAT_METADATA_JSON_BYTES:
        raise ValueError(
            f"BlobAsset metadata exceeds {MAX_FORMAT_METADATA_JSON_BYTES} canonical JSON bytes"
        )

    return BlobAssetEnvelope(
        data=payload,
        media_type=cast(str | None, media_type),
        filename=cast(str | None, filename),
        metadata=metadata_object,
    )


__all__ = ["BlobAssetEnvelope", "decode_blob_asset"]
